File: src/solver.rs
use std::collections::HashMap;
use log::warn;

use crate::parse::{CourseListContext, SectionMeeting, MeetingType};

#[derive(Debug)]
pub struct CoursePreferences {
    // 0, 1, 2, 3 are lecture ids for any 4 classes
    lecture_ids: Vec<u64>, 

    // indexed from 0 to whatever as 'internalized' idx for faster lookup and bitmasking
    sections: Vec<SectionMeeting>,

    // grouping by discussion and lab sections sections_idx indexed by lecture_ids' idx
    lab_sections: Vec<Vec<usize>>,
    lecture_sections: Vec<Vec<usize>>,
    discussion_sections: Vec<Vec<usize>>,
    sentinel_idx: usize,

    exam_sections: Vec<Option<usize>>
}

pub struct BTSolver {
    prefs: CoursePreferences,
}

impl CoursePreferences {
    pub fn new(lecture_ids: Vec<u64>, course_ctx: CourseListContext) -> Self {
        let mut lab_sections = vec![Vec::new(); lecture_ids.len()];
        let mut discussion_sections = vec![Vec::new(); lecture_ids.len()];
        let mut lecture_sections = vec![Vec::new(); lecture_ids.len()];
        let mut exam_sections = vec![None; lecture_ids.len()];

        let mut lecture_id_to_idx: HashMap<u64, usize> = HashMap::new();
        lecture_ids.iter().enumerate().for_each(|(i, k)| {
            lecture_id_to_idx.insert(*k, i);
        });

        let mut sections = course_ctx.meetings_from_lectures(&lecture_ids);
        sections.sort_by_key(|section| section.u_start);

        sections.iter().enumerate().for_each(|(idx, section)| {
            let lecture_idx = lecture_id_to_idx[&section.lecture_id];

            match section.meeting_type {
                MeetingType::Lecture => lecture_sections[lecture_idx].push(idx),
                MeetingType::Discussion => discussion_sections[lecture_idx].push(idx),
                MeetingType::Lab => lab_sections[lecture_idx].push(idx),
                MeetingType::Exam => exam_sections[lecture_idx] = Some(idx),
                MeetingType::Other => warn!("building course preferences index OTHER: {:?}", section),
            };
        });

        let sentinel_idx = sections.len();
        // this is for courses without lab or discussion or exam. just give a sentinel idx past sections' size 
        for lecture_idx in 0..lecture_ids.len() {
            if lab_sections[lecture_idx].len() == 0 {
                lab_sections[lecture_idx].push(sentinel_idx);
            }
    
            if discussion_sections[lecture_idx].len() == 0 {
                discussion_sections[lecture_idx].push(sentinel_idx);
            }
        }

        Self {
            lecture_ids,
            sections,

            lab_sections,
            lecture_sections,
            discussion_sections,
            sentinel_idx,

            exam_sections
        }
    }
}

fn mask_values(indices: &Vec<usize>, schedule_mask: &mut Vec<bool>, val: bool) {
    indices.iter().for_each(|&idx| {
        schedule_mask[idx] = val;
    });
}

impl BTSolver {
    pub fn new(prefs: CoursePreferences) -> Self {
        BTSolver {
            prefs
        }
    }

    pub fn solve(&self) -> Vec<Vec<SectionMeeting>> {
        let mut schedule_mask = vec![false; self.prefs.sections.len()];
        let mut solutions = Vec::new();

        self.search(&mut solutions, 0, 0, &mut schedule_mask);
        
        solutions
    }

    fn search(&self, solutions: &mut Vec<Vec<SectionMeeting>>, n_added: usize, n_considered: usize, schedule_mask: &mut Vec<bool>) -> usize {
        if n_added == self.prefs.lecture_ids.len() { 
            let mut valid_sections = Vec::new();
            schedule_mask.iter().enumerate().for_each(|(idx, &val)| {
                if val {
                    valid_sections.push(self.prefs.sections[idx].clone());
                }
            });

            solutions.push(valid_sections);
            return n_added; 
        }
        
        if n_considered == self.prefs.lecture_ids.len() { return n_added; }

        // add the class and add all of its lecture sections
        mask_values(&self.prefs.lecture_sections[n_added], schedule_mask, true);
        // exam section
        if let Some(exam_idx) = self.prefs.exam_sections[n_added] {
            schedule_mask[exam_idx] = true;
        }

        for lab_idx in &self.prefs.lab_sections[n_added] {
            // purpose of sentinel index is to let classes without lab or discussion be considered
            if *lab_idx != self.prefs.sentinel_idx {
                // take the lab
                schedule_mask[*lab_idx] = true;
                if self.score(&schedule_mask) < 0.0 { // check early instead of backtrack
                    schedule_mask[*lab_idx] = false;
                    continue; 
                }
            }

            for discussion_idx in &self.prefs.discussion_sections[n_added] {
                if *discussion_idx != self.prefs.sentinel_idx {
                    schedule_mask[*discussion_idx] = true;
                    if self.score(&schedule_mask) < 0.0 { // check early 
                        schedule_mask[*discussion_idx] = false;
                        continue; 
                    }
                }

                // recurse
                self.search(solutions, n_added+1, n_considered+1, schedule_mask);

                if *discussion_idx != self.prefs.sentinel_idx { schedule_mask[*discussion_idx] = false; }
            }

            if *lab_idx != self.prefs.sentinel_idx { schedule_mask[*lab_idx] = false; }
        }

        // or don't add it
        mask_values(&self.prefs.lecture_sections[n_added], schedule_mask, false);
        if let Some(exam_idx) = self.prefs.exam_sections[n_added] {
            schedule_mask[exam_idx] = false;
        }

        // // recurse, missing one course
        self.search(solutions, n_added, n_considered+1, schedule_mask);
    

        return n_added;
    }

    // assume schedule is always sorted by start date
    // the purpose of returning a float is to allow incentives for user preferences
    // probably will add a feature like this if this scheduler actually gets used
    fn score(&self, schedule_mask: &Vec<bool>) -> f64 {
        let mut ending = 0; // pretty sure no classes start sunday 12am
        for section_idx in 0..schedule_mask.len() {
            if !schedule_mask[section_idx] { continue; }
            let section = &self.prefs.sections[section_idx];
            if section.u_start <= ending { return -1.0; }

            ending = section.u_end;
        }

        1.0
    }
}

#[cfg(test)]
mod tests {
    use std::fs;
    use crate::parse::CourseListContext;
    use crate::solver::{BTSolver, CoursePreferences};
    
    #[test]
    fn score_1_class() {
        let res = fs::read("data/mess.json");
        let gql_response = serde_json::from_str(std::str::from_utf8(&res.unwrap()).unwrap()).unwrap();
        let want = vec![2023337427];
        let ctx = CourseListContext::new(&gql_response);
        let prefs = CoursePreferences::new(want, ctx);
        let solver = BTSolver::new(prefs);
        let res = solver.solve();
        res.iter().for_each(|v| { println!("{:?}", v) });
        // println!("{:?}", v["data"]["classes"]["nodes"].as_array().unwrap().len())
    }

    #[test]
    fn score_my_schedule() {
        let res = fs::read("data/mess.json");
        let gql_response = serde_json::from_str(std::str::from_utf8(&res.unwrap()).unwrap()).unwrap();
        let want = vec![2023337427];
        let ctx = CourseListContext::new(&gql_response);
        let prefs = CoursePreferences::new(want, ctx);
        let solver = BTSolver::new(prefs);
        let res = solver.solve();
        res.iter().for_each(|v| { println!("{:?}", v) });
        // println!("{:?}", v["data"]["classes"]["nodes"].as_array().unwrap().len())
    }

    #[test]
    fn my_schedule_test() {
        let res = fs::read("data/mess.json");
        let gql_response = serde_json::from_str(std::str::from_utf8(&res.unwrap()).unwrap()).unwrap();
        let want = vec![2023337427, 2023337795,  2023336415, 2023337412];
        let ctx = CourseListContext::new(&gql_response);
        let prefs = CoursePreferences::new(want, ctx);
        // prefs.sections.iter().for_each(|a| { println!("{:?}", a) });
        let solver = BTSolver::new(prefs);
        let res = solver.solve();
        res.iter().for_each(|v| { 
            v.iter().for_each(|a| { println!("{:?}", a) });
            println!();
        });
        // println!("{:?}", v["data"]["classes"]["nodes"].as_array().unwrap().len())
    }

    // #[test]
    // fn score_ess001_fail() {
    //     let res = fs::read("data/mess.json");
    //     let gql_response = serde_json::from_str(std::str::from_utf8(&res.unwrap()).unwrap()).unwrap();
    //     let want = vec![2023333322];
    //     todo!("fix this");
    //     let ctx = CourseListContext::new(&gql_response);
    //     let prefs = CoursePreferences::new(want, ctx);
    //     let solver = BTSolver::new(prefs);
    //     let res = solver.solve();
    //     res.iter().for_each(|v| { println!("{:?}", v) });
    //     // println!("{:?}", v["data"]["classes"]["nodes"].as_array().unwrap().len())
    // }
}

File: scripts/solver.py
from collections import namedtuple, defaultdict

fields = ["u_start", "u_end", "section_id", "lecture_id", "section_name", "meeting_type"]
SectionMeeting = namedtuple("SectionMeeting", fields)

data = [[135000, 139500, 2023337427, 2023337427, 'CSE-185-01', 'Lecture'], [140400, 144900, 2023337412, 2023337412, 'CSE-150-01', 'Lecture'], [226800, 231300, 2023336415, 2023336415, 'CSE-168-01', 'Lecture'], [237600, 247500, 2023337795, 2023337795, 'CSE-175-01', 'Lecture'], [243000, 253200, 2023337413, 2023337412, 'CSE-150-03L', 'Lab'], [297000, 307200, 2023336684, 2023336415, 'CSE-168-03L', 'Lab'], [307800, 312300, 2023337427, 2023337427, 'CSE-185-01', 'Lecture'], [313200, 317700, 2023337412, 2023337412, 'CSE-150-01', 'Lecture'], [329400, 339600, 2023337429, 2023337427, 'CSE-185-04L', 'Lab'], [399600, 404100, 2023336415, 2023336415, 'CSE-168-01', 'Lecture'], [459000, 468900, 2023337796, 2023337795, 'CSE-175-02L', 'Lab'], [469800, 479700, 2023337797, 2023337795, 'CSE-175-03L', 'Lab'], [480600, 490500, 2023337798, 2023337795, 'CSE-175-04L', 'Lab'], [491400, 501300, 2023337799, 2023337795, 'CSE-175-05L', 'Lab'], [491400, 501600, 2023337414, 2023337412, 'CSE-150-04L', 'Lab'], [892800, 903600, 2023337795, 2023337795, 'CSE-175-01', 'Exam'], [905400, 916200, 2023336415, 2023336415, 'CSE-168-01', 'Exam'], [991800, 1002600, 2023337412, 2023337412, 'CSE-150-01', 'Exam'], [1164600, 1175400, 2023337427, 2023337427, 'CSE-185-01', 'Exam']]

sections = [SectionMeeting(*x) for x in data]

# make sections into lists
labs = defaultdict(list)
lectures = defaultdict(list)
discussions = defaultdict(list)
exams = defaultdict(list)

for section in sections:
    if section.meeting_type == "Lecture":
        lectures[section.lecture_id].append(section)
    elif section.meeting_type == "Lab":
        labs[section.lecture_id].append(section)
    elif section.meeting_type == "Discussion":
        discussions[section.lecture_id].append(section)
    elif section.meeting_type == "Exam":
        exams[section.lecture_id].append(section)

all_courses = [*set(section.lecture_id for section in sections)]

def is_valid(schedule):
    schedule_sorted = sorted(list(schedule), key=lambda x: x.u_start)
    for i in range(1, len(schedule_sorted)):
        if schedule_sorted[i].u_start <= schedule_sorted[i-1].u_end:
            return False
    return True

def add_lectures(target_set, lectures):
    for lecture in lectures:
        target_set.add(lecture)

def remove_lectures(target_set, lectures):
    for lecture in lectures:
        target_set.remove(lecture)

schedule = set()
classes = []
def solve(consider_idx):
    if not is_valid(schedule):
        return
    
    if len(classes) == len(all_courses): 
        for section in schedule:
            print(section)
        print()
        return
    
    if consider_idx == len(all_courses): # not max # of classes
        return
    
    # take the class
    lecture_id = all_courses[consider_idx]
    lecture_meetings = lectures[lecture_id]
    classes.append(lecture_id)
    add_lectures(schedule, lecture_meetings)

    # then pick a lab
    for lab in labs[lecture_id]:
        schedule.add(lab)
        # then pick an exam
        for exam in exams[lecture_id]:
            schedule.add(exam)
            solve(consider_idx+1)
            schedule.remove(exam)
            
        schedule.remove(lab)
    
    # or not take the class
    classes.pop()
    remove_lectures(schedule, lecture_meetings)
    solve(consider_idx+1)

solve(0)
